src/trace/log.zig

```diff
@@ -171,79 +169,55 @@ pub const ChannelPrintLogger = struct {
-        }
-        return .noop;
-    }
-
-    pub fn debug(self: *Self, comptime maybe_scope: ?[]const u8) Entry {
-        if (@intFromEnum(self.max_level) >= @intFromEnum(Level.debug)) {
-            return Entry{ .channel_print = ChannelPrintEntry.init(self.log_allocator, maybe_scope, self.channel, Level.debug) };
-        }
-        return .noop;
-    }
-
-    pub fn trace(self: *Self, comptime maybe_scope: ?[]const u8) Entry {
-        if (@intFromEnum(self.max_level) >= @intFromEnum(Level.trace)) {
-            return Entry{ .channel_print = ChannelPrintEntry.init(self.log_allocator, maybe_scope, self.channel, Level.trace) };
-        }
-        return .noop;
-    }
+    pub fn log(
+        self: *Self,
+        comptime scope: ?[]const u8,
+        level: Level,
+        fields: anytype,
+        comptime fmt: []const u8,
+        args: anytype,
+    ) void {
+        if (@intFromEnum(self.max_level) < @intFromEnum(level)) return;
+        const size = logfmt.countLog(scope, level, fields, fmt, args);
+        const msg_buf = self.allocBuf(size) catch |err| {
+            std.debug.print("allocBuff failed with err: {any}", .{err});
+            return;
+        };
+        var stream = std.io.fixedBufferStream(msg_buf);
+        logfmt.writeLog(stream.writer(), scope, level, fields, fmt, args) catch |err| {
+            std.debug.print("writeLog failed with err: {any}", .{err});
+            self.log_allocator.free(msg_buf);
+            return;
+        };
+        std.debug.assert(size == stream.pos);
 
-    pub fn log(self: *Self, comptime scope: ?[]const u8, level: Level, comptime message: []const u8) void {
-        if (@intFromEnum(self.max_level) >= @intFromEnum(level)) {
-            switch (level) {
-                .err => self.err(scope).log(message),
-                .warn => self.warn(scope).log(message),
-                .info => self.info(scope).log(message),
-                .debug => self.debug(scope).log(message),
-                .trace => self.trace(scope).log(message),
-            }
-        }
+        self.channel.send(msg_buf) catch |err| {
+            std.debug.print("Send msg through channel failed with err: {any}", .{err});
+            self.log_allocator.free(msg_buf);
+            return;
+        };
     }
 
-    pub fn logf(self: *Self, comptime scope: ?[]const u8, level: Level, comptime fmt: []const u8, args: anytype) void {
-        if (@intFromEnum(self.max_level) >= @intFromEnum(level)) {
-            switch (level) {
-                .err => self.err(scope).logf(fmt, args),
-                .warn => self.warn(scope).logf(fmt, args),
-                .info => self.info(scope).logf(fmt, args),
-                .debug => self.debug(scope).logf(fmt, args),
-                .trace => self.trace(scope).logf(fmt, args),
-            }
+    // Utility function for allocating memory from RecycleFBA for part of the log message.
+    fn allocBuf(self: *const Self, size: u64) ![]u8 {
```

Comment:
Could you please explain the purpose of this function? Can `log_allocator.alloc` spuriously fail and so we need to call it multiple times? I'm pretty confused why this exists.

Reply:
I don't like it either, just copied the existing code to keep the behavior the same.

The basic idea is that you may need to wait for the RecycleFBA allocator to have some memory available. Personally I don't think we should rely on a mechanism like this, but I wasn't trying to deal with it in this pr.

Reply:
@0xNineteen what do you think? Do we need this? How about we just provide enough memory so we don't need to worry about this?

Reply:
this way provides a way to recover from no memory available which you could eventually recover from -- not sure how we could do it better -- if you want to provide enough memory so we don't need to worry about this, you can increasing the buffer size and panic on fail instead of retrying?

Reply:
> this way provides a way to recover from no memory available which you could eventually recover from

I believe this idea in and of itself is a reasonable goal, I think the main issue is that it's being done through the `std.mem.Allocator` interface, which does not provide any guarantee that this is the behavior it will display, and it's a fairly unconventional pattern as opposed to something more common such as "`arena`-like behavior".

A simple direct improvement imo would be to directly use `log_allocator_state` instead of going through the `log_allocator` field, encoding the intention for this behavior right in the code itself.

Reply:
> feels like changes are outside the scope of this PR though

https://github.com/orgs/Syndica/projects/2/views/10?pane=issue&itemId=83096299